File: bindings/junit/MagnetoAssertions.java
package com.magnetoserge.junit;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.Assertions;

import java.io.File;
import java.io.IOException;
import java.net.http.HttpResponse;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.List;

/**
 * Custom JUnit assertions for Magnéto-Serge cassettes
 *
 * Usage:
 * <pre>
 * import static com.magnetoserge.junit.MagnetoAssertions.*;
 *
 * {@literal @}Test
 * public void testUserLogin() {
 *     HttpResponse&lt;String&gt; response = client.send(request, HttpResponse.BodyHandlers.ofString());
 *     assertMatchesCassette(response, "user-login");
 * }
 * </pre>
 */
public class MagnetoAssertions {

    private static final ObjectMapper objectMapper = new ObjectMapper();
    private static String cassetteDir = "./cassettes";

    /**
     * Set the directory where cassettes are stored
     * @param dir Cassette directory path
     */
    public static void setCassetteDirectory(String dir) {
        cassetteDir = dir;
    }

    /**
     * Load a cassette from disk
     * @param name Cassette name (without extension)
     * @return Parsed cassette as JsonNode
     * @throws IOException if cassette cannot be read
     */
    private static JsonNode loadCassette(String name) throws IOException {
        Path jsonPath = Paths.get(cassetteDir, name + ".json");
        Path msgpackPath = Paths.get(cassetteDir, name + ".msgpack");

        Path cassettePath;
        if (Files.exists(jsonPath)) {
            cassettePath = jsonPath;
        } else if (Files.exists(msgpackPath)) {
            throw new UnsupportedOperationException("MessagePack cassettes not yet supported");
        } else {
            throw new IOException("Cassette not found: " + name);
        }

        String content = Files.readString(cassettePath);
        return objectMapper.readTree(content);
    }

    /**
     * Assert that HTTP response matches a cassette
     * @param response HTTP response to validate
     * @param cassetteName Name of cassette to match against
     */
    public static void assertMatchesCassette(HttpResponse<?> response, String cassetteName) {
        try {
            JsonNode cassette = loadCassette(cassetteName);
            String method = response.request().method();
            String url = response.request().uri().toString();

            JsonNode interaction = findMatchingInteraction(cassette, method, url);

            if (interaction == null) {
                Assertions.fail(String.format(
                    "No matching interaction found in cassette '%s' for %s %s",
                    cassetteName, method, url
                ));
            }

            // Validate status code
            int expectedStatus = interaction.get("kind").get("Http").get("response").get("status").asInt();
            int actualStatus = response.statusCode();

            Assertions.assertEquals(expectedStatus, actualStatus,
                String.format("Status code mismatch in cassette '%s'", cassetteName));

        } catch (IOException e) {
            Assertions.fail("Failed to load cassette: " + e.getMessage());
        }
    }

    /**
     * Assert that HTTP response status matches cassette
     * @param response HTTP response to validate
     * @param cassetteName Name of cassette
     * @param expectedStatus Expected status code
     */
    public static void assertCassetteStatus(HttpResponse<?> response, String cassetteName, int expectedStatus) {
        try {
            JsonNode cassette = loadCassette(cassetteName);
            String method = response.request().method();
            String url = response.request().uri().toString();

            JsonNode interaction = findMatchingInteraction(cassette, method, url);

            if (interaction == null) {
                Assertions.fail(String.format(
                    "No matching interaction found in cassette '%s'",
                    cassetteName
                ));
            }

            int cassetteStatus = interaction.get("kind").get("Http").get("response").get("status").asInt();

            Assertions.assertEquals(expectedStatus, cassetteStatus,
                String.format("Expected status %d but cassette has %d", expectedStatus, cassetteStatus));

        } catch (IOException e) {
            Assertions.fail("Failed to load cassette: " + e.getMessage());
        }
    }

    /**
     * Assert that cassette has expected number of interactions
     * @param cassetteName Name of cassette
     * @param expectedCount Expected interaction count
     */
    public static void assertInteractionCount(String cassetteName, int expectedCount) {
        try {
            JsonNode cassette = loadCassette(cassetteName);
            int actualCount = cassette.get("interactions").size();

            Assertions.assertEquals(expectedCount, actualCount,
                String.format("Expected %d interactions but found %d", expectedCount, actualCount));

        } catch (IOException e) {
            Assertions.fail("Failed to load cassette: " + e.getMessage());
        }
    }

    /**
     * Assert that cassette contains cookies
     * @param cassetteName Name of cassette
     */
    public static void assertHasCookies(String cassetteName) {
        try {
            JsonNode cassette = loadCassette(cassetteName);
            JsonNode cookies = cassette.get("cookies");

            Assertions.assertNotNull(cookies, "Cassette has no cookies");
            Assertions.assertTrue(cookies.size() > 0,
                "Cassette cookies array is empty");

        } catch (IOException e) {
            Assertions.fail("Failed to load cassette: " + e.getMessage());
        }
    }

    /**
     * Assert that cassette has specific cookie
     * @param cassetteName Name of cassette
     * @param cookieName Name of cookie to find
     */
    public static void assertHasCookie(String cassetteName, String cookieName) {
        try {
            JsonNode cassette = loadCassette(cassetteName);
            JsonNode cookies = cassette.get("cookies");

            if (cookies == null || cookies.size() == 0) {
                Assertions.fail("Cassette has no cookies");
            }

            boolean found = false;
            for (JsonNode cookie : cookies) {
                if (cookie.get("name").asText().equals(cookieName)) {
                    found = true;
                    break;
                }
            }

            Assertions.assertTrue(found,
                String.format("Cookie '%s' not found in cassette", cookieName));

        } catch (IOException e) {
            Assertions.fail("Failed to load cassette: " + e.getMessage());
        }
    }

    /**
     * Assert that cassette version matches expected
     * @param cassetteName Name of cassette
     * @param expectedVersion Expected version (e.g., "2.0")
     */
    public static void assertCassetteVersion(String cassetteName, String expectedVersion) {
        try {
            JsonNode cassette = loadCassette(cassetteName);
            String actualVersion = cassette.get("version").asText();

            Assertions.assertEquals(expectedVersion, actualVersion,
                String.format("Expected version %s but found %s", expectedVersion, actualVersion));

        } catch (IOException e) {
            Assertions.fail("Failed to load cassette: " + e.getMessage());
        }
    }

    /**
     * Find matching interaction in cassette
     * @param cassette Loaded cassette
     * @param method HTTP method
     * @param url Request URL
     * @return Matching interaction or null
     */
    private static JsonNode findMatchingInteraction(JsonNode cassette, String method, String url) {
        JsonNode interactions = cassette.get("interactions");

        for (JsonNode interaction : interactions) {
            if (!interaction.has("kind") || !interaction.get("kind").has("Http")) {
                continue;
            }

            JsonNode httpInteraction = interaction.get("kind").get("Http");
            JsonNode request = httpInteraction.get("request");

            String reqMethod = request.get("method").asText();
            String reqUrl = request.get("url").asText();

            if (reqMethod.equals(method) && reqUrl.equals(url)) {
                return interaction;
            }
        }

        return null;
    }
}

File: bindings/phpunit/MagnetoAssertions.php
<?php

namespace MagnetoSerge\PHPUnit;

use PHPUnit\Framework\Assert;
use PHPUnit\Framework\AssertionFailedError;

/**
 * Custom PHPUnit assertions for Magnéto-Serge cassettes
 *
 * Usage:
 * <code>
 * use MagnetoSerge\PHPUnit\MagnetoAssertions;
 *
 * class UserTest extends TestCase
 * {
 *     use MagnetoAssertions;
 *
 *     public function testUserLogin()
 *     {
 *         $response = $this->client->post('/api/authenticate', [...]);
 *         $this->assertMatchesCassette($response, 'user-login');
 *     }
 * }
 * </code>
 */
trait MagnetoAssertions
{
    /**
     * Cassette directory path
     * @var string
     */
    protected static $cassetteDir = './cassettes';

    /**
     * Set the cassette directory
     *
     * @param string $dir Directory path
     * @return void
     */
    public static function setCassetteDirectory(string $dir): void
    {
        self::$cassetteDir = $dir;
    }

    /**
     * Load a cassette from disk
     *
     * @param string $name Cassette name (without extension)
     * @return array Parsed cassette
     * @throws \RuntimeException if cassette not found
     */
    protected function loadCassette(string $name): array
    {
        $jsonPath = self::$cassetteDir . '/' . $name . '.json';
        $msgpackPath = self::$cassetteDir . '/' . $name . '.msgpack';

        if (file_exists($jsonPath)) {
            $content = file_get_contents($jsonPath);
            return json_decode($content, true);
        } elseif (file_exists($msgpackPath)) {
            throw new \RuntimeException('MessagePack cassettes not yet supported');
        } else {
            throw new \RuntimeException("Cassette not found: {$name}");
        }
    }

    /**
     * Find matching interaction in cassette
     *
     * @param array $cassette Loaded cassette
     * @param string $method HTTP method
     * @param string $url Request URL
     * @return array|null Matching interaction or null
     */
    protected function findMatchingInteraction(array $cassette, string $method, string $url): ?array
    {
        foreach ($cassette['interactions'] as $interaction) {
            if (!isset($interaction['kind']['Http'])) {
                continue;
            }

            $request = $interaction['kind']['Http']['request'];

            if ($request['method'] === $method && $request['url'] === $url) {
                return $interaction;
            }
        }

        return null;
    }

    /**
     * Assert that HTTP response matches a cassette
     *
     * @param \Psr\Http\Message\ResponseInterface|array $response HTTP response
     * @param string $cassetteName Name of cassette to match against
     * @return void
     * @throws AssertionFailedError
     */
    public function assertMatchesCassette($response, string $cassetteName): void
    {
        $cassette = $this->loadCassette($cassetteName);

        // Extract request info (support different HTTP client interfaces)
        if (is_array($response)) {
            $method = $response['method'] ?? 'GET';
            $url = $response['url'] ?? '';
            $status = $response['status'] ?? 0;
        } else {
            // PSR-7 Response
            $request = $response->getRequest();
            $method = $request->getMethod();
            $url = (string) $request->getUri();
            $status = $response->getStatusCode();
        }

        $interaction = $this->findMatchingInteraction($cassette, $method, $url);

        Assert::assertNotNull(
            $interaction,
            "No matching interaction found in cassette '{$cassetteName}' for {$method} {$url}"
        );

        $expectedStatus = $interaction['kind']['Http']['response']['status'];

        Assert::assertEquals(
            $expectedStatus,
            $status,
            "Status code mismatch in cassette '{$cassetteName}'"
        );
    }

    /**
     * Assert that HTTP response status matches cassette
     *
     * @param \Psr\Http\Message\ResponseInterface|array $response HTTP response
     * @param string $cassetteName Name of cassette
     * @param int $expectedStatus Expected status code
     * @return void
     * @throws AssertionFailedError
     */
    public function assertCassetteStatus($response, string $cassetteName, int $expectedStatus): void
    {
        $cassette = $this->loadCassette($cassetteName);

        // Extract request info
        if (is_array($response)) {
            $method = $response['method'] ?? 'GET';
            $url = $response['url'] ?? '';
        } else {
            $request = $response->getRequest();
            $method = $request->getMethod();
            $url = (string) $request->getUri();
        }

        $interaction = $this->findMatchingInteraction($cassette, $method, $url);

        Assert::assertNotNull(
            $interaction,
            "No matching interaction found in cassette '{$cassetteName}'"
        );

        $cassetteStatus = $interaction['kind']['Http']['response']['status'];

        Assert::assertEquals(
            $expectedStatus,
            $cassetteStatus,
            "Expected status {$expectedStatus} but cassette has {$cassetteStatus}"
        );
    }

    /**
     * Assert that cassette has expected number of interactions
     *
     * @param string $cassetteName Name of cassette
     * @param int $expectedCount Expected interaction count
     * @return void
     * @throws AssertionFailedError
     */
    public function assertInteractionCount(string $cassetteName, int $expectedCount): void
    {
        $cassette = $this->loadCassette($cassetteName);
        $actualCount = count($cassette['interactions']);

        Assert::assertEquals(
            $expectedCount,
            $actualCount,
            "Expected {$expectedCount} interactions but found {$actualCount}"
        );
    }

    /**
     * Assert that cassette contains cookies
     *
     * @param string $cassetteName Name of cassette
     * @return void
     * @throws AssertionFailedError
     */
    public function assertHasCookies(string $cassetteName): void
    {
        $cassette = $this->loadCassette($cassetteName);

        Assert::assertArrayHasKey('cookies', $cassette, 'Cassette has no cookies');
        Assert::assertNotEmpty($cassette['cookies'], 'Cassette cookies array is empty');
    }

    /**
     * Assert that cassette has specific cookie
     *
     * @param string $cassetteName Name of cassette
     * @param string $cookieName Name of cookie to find
     * @return void
     * @throws AssertionFailedError
     */
    public function assertHasCookie(string $cassetteName, string $cookieName): void
    {
        $cassette = $this->loadCassette($cassetteName);

        Assert::assertArrayHasKey('cookies', $cassette, 'Cassette has no cookies');

        $cookieNames = array_column($cassette['cookies'], 'name');

        Assert::assertContains(
            $cookieName,
            $cookieNames,
            "Cookie '{$cookieName}' not found in cassette"
        );
    }

    /**
     * Assert that cassette version matches expected
     *
     * @param string $cassetteName Name of cassette
     * @param string $expectedVersion Expected version (e.g., "2.0")
     * @return void
     * @throws AssertionFailedError
     */
    public function assertCassetteVersion(string $cassetteName, string $expectedVersion): void
    {
        $cassette = $this->loadCassette($cassetteName);
        $actualVersion = $cassette['version'] ?? null;

        Assert::assertEquals(
            $expectedVersion,
            $actualVersion,
            "Expected version {$expectedVersion} but found {$actualVersion}"
        );
    }
}
